### DISCUSSION TITLE
Extract a thumbnail

### DISCUSSION THREAD

**Discussion by xmodpt** [sn4k3/UVtools#823] [Q&A]
Ola Tiago

é possivel com o uvtools e o lychee command gerar e salvar um thumb do file em questão?

obrigado
Nuno Pires

Hello Tiago

i has woundering if there is a command line for etracting the thumbnail from a ctb and save it.

or even better a python script to do it.... do you know of ny ?

thank you
Nuno Pires

**Comment by sn4k3** [comment]
You can use command line with [UVtoolsCmd](https://github.com/sn4k3/UVtools?tab=readme-ov-file#command-line) with `extract` command


```
C:\Program Files\UVtools>UVtoolsCmd extract

Description:
  Extract file contents to a folder

Usage:
  UVtoolsCmd extract [<input-file> [<output-folder>]] [options]

Arguments:
  <input-file>     Input file to open and read
  <output-folder>  Output folder []

Options:
  --no-overwrite                          If the output folder exists do not overwrite
  -c, --content <File|Layers|Thumbnails>  Set the type of content to extract [default: File]
  -i, --index <index>                     Sets the thumbnail or layer index to extract
  -r, --range <range>                     Sets the layer range to extract
  -?, -h, --help                          Show help and usage information
```

Example (Extract all): `UVtoolsCmd extract -c Thumbnails "yourfile"`
Example (Extract index) to same file folder: `UVtoolsCmd extract -c Thumbnails -i 0 "yourfile" .`

**Reply by xmodpt** [reply]
thank you for the reply

is there a way to implement it but in python ? in order to use it in octoprint ?

regards

**Reply by sn4k3** [reply]
You can always call UVtoolsCmd from python.
You can also use UVtools Core in [python](https://github.com/sn4k3/UVtools/tree/master/Scripts/UVtools.Python), however it's more difficult.

**Reply by xmodpt** [reply]
my idea it to TRY and make a plugin for octoprint that allow the user to see thmb's for  SLA files

i kbow there is allready a plugin that does that but is only for gcode

**Reply by sn4k3** [reply]
Calling UVtoolsCmd with subprocess or system is the easiest way.
However if you only want to show thumbnails you can fetch them directly from files, it's not hard since they are in "static place" and mostly rgb 16bit

**Reply by xmodpt** [reply]
ok but my issue is how to do that in python... ( o really don't know and i am starting to get my head arround it)
i have tryed to extract the thumb using python but up so far, i have been unable to do it... it seems that ctb files are encrypted and i can find info on how to decrypt them....

np

**Comment by Hazoooka** [comment]
Is it possible the other way, to replace the thumbnail?

**Reply by sn4k3** [reply]
Yes:

```
UVtoolsCmd set-thumbnail

Description:
  Sets and replace thumbnail(s) in the file

Usage:
  UVtoolsCmd set-thumbnail [<input-file> [<file path|layer index|:random-layer|:heatmap>]] [options]

Arguments:
  <input-file>                                    Input file to open and read
  <file path|layer index|:random-layer|:heatmap>  Choose from a file, layer index, random layer or generate a heatmap
                                                  [default: :heatmap]

Options:
  -i, --indexes <indexes>  Replaces only the matching thumbnail(s) index(es)
  -?, -h, --help           Show help and usage information
```

Example: `UVtoolsCmd set-thumbnail "your-sliced-file" "thumbnail-path.png"`

**Comment by Hazoooka** [comment]
Thank you :)No better time to work than Sunday evening ;)
